docs/configuration.md
# Configuration

`dbt-score` can be configured through a `pyproject.toml` file or via the command
line.

## pyproject.toml

It is recommended to place the file in the root of your dbt project. `dbt-score`
will look for `pyproject.toml` in the directory from which it is run and its
parent directories.

An example of a `pyproject.toml` file to configure `dbt-score` can be found
below:

```toml
[tool.dbt-score]
rule_namespaces = ["dbt_score.rules", "dbt_score_rules", "custom_rules"]
disabled_rules = ["dbt_score.rules.generic.columns_have_description"]
inject_cwd_in_python_path = true
fail_project_under = 7.5
fail_any_item_under = 8.0

[tool.dbt-score.badges]
first.threshold = 10.0
first.icon = "🥇"
second.threshold = 8.0
second.icon = "🥈"
third.threshold = 6.0
third.icon = "🥉"
wip.icon = "🏗️"

[tool.dbt-score.rules."dbt_score.rules.generic.sql_has_reasonable_number_of_lines"]
severity = 1
max_lines = 250
```

### Configuration options

The following options can be set in the `pyproject.toml` file:

#### Main configuration

```toml
[tool.dbt-score]
```

- `rule_namespaces`: A list of Python namespaces to search for rules. The
  default is `["dbt_score.rules", "dbt_score_rules"]`. Be aware when overriding
  this setting, that the default rules are in `dbt_score.rules` and are disabled
  if not included here.
- `disabled_rules`: A list of rules to disable.
- `fail_project_under` (default: `5.0`): If the project score is below this
  value the command will fail with return code 1.
- `fail_any_item_under` (default: `5.0`): If any entity scores below this value
  the command will fail with return code 1.

#### Badges configuration

```toml
[tool.dbt-score.badges]
```

Four badges can be configured: `first`, `second`, `third` and `wip`. Each badge
can be configured with the following option:

- `icon`: The icon to use for the badge. A string that will be displayed in the
  output, e.g. `🥇`.

All badges except `wip` can be configured with the following option:

- `threshold`: The threshold for the badge. A decimal number between `0.0` and
  `10.0` that will be used to compare to the score. The threshold is the minimum
  score required for a model or source to be rewarded with a certain badge.

The default values can be found in the
[BadgeConfig](reference/config.md#dbt_score.config.BadgeConfig).

#### Rule configuration

```toml
[tool.dbt-score.rules."rule_namespace.rule_name"]
```

Every rule can be configured with the following option:

- `severity`: The severity of the rule. Rules have a default severity and can be
  overridden. It's an integer with a minimum value of 1 and a maximum value
  of 4.
- `rule_filter_names`: Filters used by the rule. Takes a list of names that can
  be found in the same namespace as the rules (see
  [Package rules](package_rules.md)).

  Example: the generic rule `has_example_sql` shall apply only to models
  materializing a table.

  ```toml
  [tool.dbt-score.rules."dbt_score.rules.generic.has_example_sql"]
  rule_filter_names=["dbt_score.rules.filters.is_table"]
  ```

Some rules have additional configuration options, e.g.
[sql_has_reasonable_number_of_lines](rules/generic.md#sql_has_reasonable_number_of_lines).
Depending on the rule, the options will have different names, types and default
values. In the case of the
[sql_has_reasonable_number_of_lines](rules/generic.md#sql_has_reasonable_number_of_lines),
the `max_lines` option can be configured.

## Command line

Many configuration options can also be set via the command line. To understand
how to configure `dbt-score` from the command line:

```bash
dbt-score lint --help
```


docs/reference/config.md
# Config

::: dbt_score.config


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

- Expose column config field from the manifest.

## [0.15.0] - 2025-11-19

- Round score of individual entities when showing summary.
- Support linting dbt macros as a new evaluable entity type.

## [0.14.1] - 2025-10-09

- Migrate to `uv` project manager.

## [0.14.0] - 2025-08-08

- Display the parse error message when `dbt parse` fails.
- **Breaking**: JSON-formatted output is using `unique_id` as key instead of
  `name`, to avoid duplicates (e.g if exposure and model have the same name).

## [0.13.1] - 2025-07-29

- Fix filters being applied to wrong evaluables (#124)

## [0.13.0] - 2025-07-07

- Support linting of exposures (#112)
- Fix retrieval of data_tests when using quoted columns in BigQuery (#122)

## [0.12.0] - 2025-05-06

- Add support for linting and scoring dbt seeds (#110)
- Add `children` to models, snapshots, and sources. (#113)
- Add `parents` to models and snapshots, allowing access to parent nodes. (#109)

## [0.11.0] - 2025-04-04

- Improve documentation on rule filters. (#93)
- Add `group` to model definition. (#99)
- Add rule: prevent use of is_incremental() in non-incremental models. (#103)
- Support linting of snapshots. (#96)

## [0.10.0] - 2025-01-27

- Add debug mode to help writing new rules. (#91)
- Fix tests without metadata. (#88)
- Add new rule to enforce presence of uniqueness test. (#90)
- Add new rule to enforce single-column PK to be defined at column level. (#90)
- Add new rule to enforce 1 column uniqueness test to be defined at column
  level. (#90)
- Add `constraints` to the model schema. (#90)

## [0.9.0] - 2024-12-19

- Documenting support for python 3.13. (#86)
- Only show failing rules per default in `HumanReadableFormatter`. Also added
  `--show` parameter in the CLI to change this behavior. (#77)
- Ignore imported rules and filters when building the rule registry. (#87)

## [0.8.0] - 2024-11-12

- Support linting of sources.
- **Breaking**: Renamed modules: `dbt_score.model_filter` becomes
  `dbt_score.rule_filter`
- **Breaking**: Renamed filter class and decorator: `@model_filter` becomes
  `@rule_filter` and `ModelFilter` becomes `RuleFilter`.
- **Breaking**: Config option `model_filter_names` becomes `rule_filter_names`.
- **Breaking**: CLI flag naming fixes: `--fail_any_model_under` becomes
  `--fail-any-item-under` and `--fail_project_under` becomes
  `--fail-project-under`.

## [0.7.1] - 2024-11-01

- Fix mkdocs.

## [0.7.0] - 2024-11-01

- **Breaking**: The rule `public_model_has_example_sql` has been renamed
  `has_example_sql` and applies by default to all models.
- **Breaking**: Remove `dbt-core` from dependencies. Since it is not mandatory
  for `dbt-score` to execute `dbt`, remove the dependency.
- **Breaking**: Stop using `MultiOption` selection type.

## [0.6.0] - 2024-08-23

- **Breaking**: Improve error handling in CLI. Log messages are written in
  stderr, and exit code is 2 in case of anything going wrong. (#73)
- Auto-round scores down to align scores and medals. (#74)

## [0.5.0] - 2024-08-15

- Add model filters to let models be ignored by certain rules.

## [0.4.0] - 2024-08-08

- Add null check before calling `project_evaluated` in the `evaluate` method to
  prevent errors when no models are found. (#64)
- Add JSON formatter for machine-readable output. (#68)

## [0.3.0] - 2024-06-20

### Added

- Add `project_fail_under` configuration.
- Add `fail_any_model_under` configuration.
- **Breaking:** default values of `5.0` for `project_fail_under` and
  `fail_any_model_under` will cause command to exit return code 1.

## [0.2.1] - 2024-06-17

### Added

- Lint the current dbt project only, not including the imported models.

## [0.2.0] - 2024-06-14

### Added

- Support Python 3.10.

## [0.1.3] - 2024-06-11

### Added

- Inject current working directory into python path by default.

## [0.1.2] - 2024-06-07

### Added

- Create contributors guide for the documentation website.
- Add Github icon and link to documentation website.

## [0.1.1] - 2024-06-03

### Added

- CLI based on Click.
- Ability to parse dbt's `manifest.json` into internal structures.
- Rule registry and rule discovery.
- Rule API, decorator-based or class-based.
- Linting and scoring functionality for dbt models.
- Configuration through `pyproject.toml`.
- Default rules in `dbt_score.rules.generic`.
- Badges for project and model evaluation.


src/dbt_score/config.py
"""This module is responsible for loading configuration."""

import logging
import sys
from dataclasses import dataclass, field, replace
from pathlib import Path
from typing import Any, Final

from dbt_score.rule import RuleConfig

if sys.version_info >= (3, 11):
    import tomllib
else:
    import tomli as tomllib

logger = logging.getLogger(__name__)

DEFAULT_CONFIG_FILE = "pyproject.toml"


@dataclass
class Badge:
    """Badge object."""

    icon: str
    threshold: float


@dataclass
class BadgeConfig:
    """Configuration for badges."""

    third: Badge = field(default_factory=lambda: Badge("🥉", 6.0))
    second: Badge = field(default_factory=lambda: Badge("🥈", 8.0))
    first: Badge = field(default_factory=lambda: Badge("🥇", 10.0))
    wip: Badge = field(default_factory=lambda: Badge("🚧", 0.0))

    def validate(self) -> None:
        """Validate the badge configuration."""
        if not (self.first.threshold > self.second.threshold > self.third.threshold):
            raise ValueError("Invalid badge thresholds.")
        if self.first.threshold > 10.0:  # noqa: PLR2004 [magic-value-comparison]
            raise ValueError("first threshold must 10.0 or lower.")
        if self.third.threshold < 0.0:
            raise ValueError("third threshold must be 0.0 or higher.")
        if self.wip.threshold != 0.0:
            raise AttributeError("wip badge cannot have a threshold configuration.")


class Config:
    """Configuration for dbt-score."""

    _main_section: Final[str] = "tool.dbt-score"
    _options: Final[list[str]] = [
        "rule_namespaces",
        "disabled_rules",
        "selected_rules",
        "inject_cwd_in_python_path",
        "fail_project_under",
        "fail_any_item_under",
        "show",
        "debug",
    ]
    _rules_section: Final[str] = "rules"
    _badges_section: Final[str] = "badges"

    def __init__(self) -> None:
        """Initialize the Config object."""
        self.rule_namespaces: list[str] = ["dbt_score.rules", "dbt_score_rules"]
        self.disabled_rules: list[str] = []
        self.selected_rules: list[str] = []
        self.inject_cwd_in_python_path = True
        self.rules_config: dict[str, RuleConfig] = {}
        self.config_file: Path | None = None
        self.badge_config: BadgeConfig = BadgeConfig()
        self.fail_project_under: float = 5.0
        self.fail_any_item_under: float = 5.0
        self.show: str = "failing-rules"
        self.debug: bool = False

    def set_option(self, option: str, value: Any) -> None:
        """Set an option in the config."""
        setattr(self, option, value)

    def _load_toml_file(self, file: str) -> None:
        """Load the options from a TOML file."""
        with open(file, "rb") as f:
            toml_data = tomllib.load(f)

        tools = toml_data.get("tool", {})
        dbt_score_config = tools.get("dbt-score", {})
        rules_config = dbt_score_config.pop(self._rules_section, {})
        badge_config = dbt_score_config.pop(self._badges_section, {})

        # Main configuration
        for option, value in dbt_score_config.items():
            if option in self._options:
                self.set_option(option, value)
            elif not isinstance(
                value, dict
            ):  # If value is a dictionary, it's another section
                logger.warning(
                    f"Option {option} in {self._main_section} not supported."
                )

        # Badge configuration
        for name, config in badge_config.items():
            try:
                default_config = getattr(self.badge_config, name)
                updated_config = replace(default_config, **config)
                setattr(self.badge_config, name, updated_config)
            except AttributeError as e:
                options = list(BadgeConfig.__annotations__.keys())
                raise AttributeError(f"Config only accepts badges: {options}.") from e
            except TypeError as e:
                options = list(Badge.__annotations__.keys())
                if name == "wip":
                    options.remove("threshold")
                raise AttributeError(
                    f"Badge {name}: config only accepts {options}."
                ) from e

        self.badge_config.validate()

        # Rule configuration
        self.rules_config = {
            name: RuleConfig.from_dict(config) for name, config in rules_config.items()
        }

    @staticmethod
    def get_config_file(directory: Path) -> Path | None:
        """Get the config file."""
        candidates = [directory]
        candidates.extend(directory.parents)
        for path in candidates:
            config_file = path / DEFAULT_CONFIG_FILE
            if config_file.exists():
                return config_file

    def load(self) -> None:
        """Load the config."""
        config_file = self.get_config_file(Path.cwd())
        if config_file:
            self._load_toml_file(str(config_file))

    def overload(self, values: dict[str, Any]) -> None:
        """Overload config with additional values."""
        for key, value in values.items():
            self.set_option(key, value)
